WordPress/openverse-frontend
# Filter modal

## Fixes


Related to #368 by @obulat

## Description


This PR adds a dynamic component for filters: one that is a full-page modal for mobile widths, but a sidebar for tablet and up. This helps us to avoid having two elements with the same ids for inputs, one hidden on mobile screens and one hidden on desktops, as we previously had. 

The VHeaderFilter component is a wrapper (with implementation mostly based on VModal component) over a VFilterButton, and the VTeleport to the sidebar on desktop, or a VModalContainer on mobile screens.

To overcome the Nuxt SSR hydration problem described in the `Additional details` section, the component is initially set to the mobile version, and is changed to the desktop version in `onMounted` hook, if necessary.

Other changes:
- `VModalContent` is updated to allow for creating the full-page mobile modals.
- The `VButton` is using `aria-pressed` and `aria-expanded` states to add the `pressed` styles.
- `tertiary` `VButton` variant is now a bordered button, both in resting and in hovered states. It is turned into `action-menu` variant, which is a bordereless button, when the header is scrolled.
- `VSearchGridFilter` is refactored to use composition API.

## Additional details
Using a `computed` property returning a dynamic component based on the screen width causes problems. By default, when server-rendered, the `isMdScreen` value is false, so the full-screen modal is rendered on the server. Then if you open the app on a desktop, the `computed` returns the sidebar teleport. This causes a SSR - CSR mismatch, and a full re-render. Sometimes, this full re-render also causes Nuxt to mount some components twice.

```
Parent:  
<button class="transition-colors self-c…Ks-oJ size-medium_faRVV" type="button" aria-controls="filter-sidebar" aria-label="0 Filters" aria-haspopup="dialog">
vue.common.dev.js:6459
Mismatching childNodes vs. VNodes:  
NodeList(3) [ svg.v-icon.w-6.h-6, #text, span ]
Array(3) [ {…}, {…}, {…} ]
vue.common.dev.js:6460
[Vue warn]: The client-side rendered virtual DOM tree is not matching server-rendered content. This is likely caused by incorrect HTML markup, for example nesting block-level elements inside <p>, or missing <tbody>. Bailing hydration and performing full client-side render. vue.common.dev.js:630
```

The same error is thrown when we render something using `v-if` based on screen width (as in when we decide whether to show an icon on a button or not).
It seems that this is a know issue in Nuxt without a solution: https://github.com/nuxt/ nuxt.js/issues/9441.
## Testing Instructions

Run `pnpm run dev`.
On any page besides the search pages, there should be no filter button in the header, and no filter sidebar, even if the it is set to open in the local storage (you can open the sidebar on the search page to set the local storage value to true, and then reload the homepage to test this).
On a search page, the header should have a filters button. You can test that it opens a sidebar on desktop, but a modal on mobile. The desktop version should change into mobile if you change screen width.

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
